src/lib/bootloader/udev_mapping.rb

```diff
@@ -76,52 +92,61 @@ def udev_to_kernel(dev)
       device.name
     end
 
-    # Udev name for the device
-    #
-    # The selected udev name depends on the mount by option. In case of an unmounted device,
-    # storage-ng has logic to discover the preferred mount by option.
-    #
-    # @see #to_mountby_device
-    #
-    # @return [String]
     def kernel_to_udev(dev)
       device = Y2Storage::BlkDevice.find_by_name(staging, dev)
       if device.nil?
         log.error "Cannot find #{dev}"
         return dev
       end
 
-      udev = udev_path(device)
-      log.info "udev device for #{dev.inspect} is #{udev.inspect}"
+      result = udev_name_for(device)
+      log.info "udev device for #{dev.inspect} is #{result.inspect}"
 
-      udev
+      result
     end
 
-    # Most convenient udev name of the device
-    #
-    # If possible, the udev path is chosen based on the mount_by attribute of
-    # the filesystem. If the device is not mounted or there is no path for
-    # the specified mount_by, a preferred (and available) name is calculated.
-    #
-    # @param device [Y2Storage::BlkDevice]
-    # @return [String]
-    def udev_path(device)
-      filesystem = device.filesystem
-
-      if filesystem
-        mount_by_name = filesystem.mount_by_name
+    # picks udev name according to strategy
+    # @see #to_mountby_device
+    def udev_name_for(device)
+      mount_by_udev(device) ||
+        udev_by_label(device) ||
+        udev_by_uuid(device) ||
+        udev_by_id(device) ||
+        udev_by_path(device) ||
+        device.name
+    end
 
-        if mount_by_name
-          log.info "udev_path: using the udev name of the configured mount_by"
-          mount_by_name
-        else
-          log.info "udev_path: using the preferred udev name for the filesystem"
-          filesystem.preferred_name
-        end
+    def mount_by_udev(device)
+      filesystem = device.filesystem
+      return nil unless filesystem
+      # mount_by is nil, so not mounted and we need to use our own strategy
+      return nil if filesystem.mount_by.nil?
+
+      case filesystem.mount_by.to_sym
+      when :device then device.name
+      when :uuid then udev_by_uuid(device)
+      when :label then udev_by_label(device)
+      when :id then udev_by_id(device)
+      when :path then udev_by_path(device)
       else
-        log.info "udev_path: not formatted, using preferred udev name for the block device"
-        device.preferred_name
+        raise "Unknown mount by option #{filesystem.mount_by.inspect} for #{filesystem.inspect}"
       end
     end
+
+    def udev_by_uuid(device)
+      device.udev_full_uuid
+    end
+
+    def udev_by_label(device)
+      device.udev_full_label
+    end
+
+    def udev_by_id(device)
+      device.udev_full_ids.first
```

Comment:
sadly this is also source of bugs, because udev does not consider all ids equal and some are better then others :(

Reply:
libstorage-ng filters many of the ids out. Only the reliable ones (following libstorage-ng criteria) are included in `#udev_full_ids`.